### PR TITLE
Make the client poll the master concurrently

### PR DESCRIPTION
**Pull request by qian0817**
### What changes are proposed in this pull request?

Make the client poll the master concurrently.

### Why are the changes needed?
Currently, we find the active master node by polling each node in series. If a node did not respond due to network or other issues, this may take a lot of time and affect service availability. Make the client poll the master concurrently can alleviate this problem.

### CODE REVIEW

**Comment by ChunxuTang** [comment]
Hi @qian0817, thanks for your contribution!

I'm a bit hesitant to accept this design. The concurrent polling may also impose an unexpected load on the master node. Have you discussed this design with any Alluxio engineers before?

**Comment by qian0817** [comment]
I have not discussed this offline with alluxio's engineers.
I don't think this will cause a load increase on the master node, because now all requests will be sent to the first node, and after the modification all requests will be sent to every node. This will only cause the load on all nodes to be the same as the load on the first node before, without raising the new load.

**Comment by ChunxuTang** [comment]
@qian0817 Let me involve some experts in this thread.

Hi @jiacheliu3, as you're an expert on the Alluxio master, would you mind taking a look at this PR? Thanks.

File: core/common/src/test/java/alluxio/conf/ConfigurationBuilder.java [hunk 1]
```diff
@@ -24,7 +24,7 @@ public class ConfigurationBuilder {
    * @return the updated configuration builder
    */
   public ConfigurationBuilder setProperty(PropertyKey key, Object value) {
-    mProperties.put(key, value.toString(), Source.RUNTIME);
+    mProperties.put(key, value, Source.RUNTIME);
```
Comment (elega) [inline core/common/src/test/java/alluxio/conf/ConfigurationBuilder.java:27]: why is this change needed?

File: core/common/src/main/java/alluxio/master/PollingMasterInquireClient.java [hunk 2]
```diff
@@ -129,31 +143,81 @@ private InetSocketAddress getAddress() {
...
+    }
+  }
+
+  @Nullable
+  private InetSocketAddress findActiveAddressConcurrent(List<InetSocketAddress> addresses) {
+    List<Future<InetSocketAddress>> futures = new ArrayList<>(addresses.size());
+    try {
+      ExecutorCompletionService<InetSocketAddress> completionService =
+          new ExecutorCompletionService<>(EXECUTOR_SERVICE);
+      for (InetSocketAddress address : addresses) {
```
Comment (elega) [inline core/common/src/main/java/alluxio/master/PollingMasterInquireClient.java:159]: though we run those three polling tasks concurrently here essentially we have to wait all of them to finish. I'm thinking if we can do better and return it immediately as soon as a validate result comes up. 

I'm thinking if we can leverage the completable future. Here are some examples:
https://stackoverflow.com/questions/36612224/in-java-how-do-i-process-completablefutures-and-get-the-first-desireable-result

File: core/common/src/main/java/alluxio/network/RejectingServer.java [hunk 3]
```diff
@@ -31,13 +32,23 @@ public final class RejectingServer extends Thread {
...
    */
   public RejectingServer(InetSocketAddress address) {
+    this(address, 0);
+  }
+
+  /**
+   * @param address the socket address to reject requests on
+   * @param sleepTime sleep time before close connection
+   */
+  public RejectingServer(InetSocketAddress address, long sleepTime) {
```
Comment (elega) [inline core/common/src/main/java/alluxio/network/RejectingServer.java:48]: is this only used in testing? if yes, add an VisibleForTesting annotation

Comment (qian0817) [reply core/common/src/main/java/alluxio/network/RejectingServer.java:50]: fix

File: core/common/src/main/java/alluxio/master/PollingMasterInquireClient.java (hunk 2 above)
Comment (qian0817) [reply core/common/src/main/java/alluxio/master/PollingMasterInquireClient.java:159]: Please see line 164. The ExecutorCompletionService#take method retrieves the first Future that has completed and, if its value is not null, returns it directly while canceling any remaining Futures at line 175.

File: core/common/src/test/java/alluxio/conf/ConfigurationBuilder.java (hunk 1 above)
Comment (qian0817) [reply core/common/src/test/java/alluxio/conf/ConfigurationBuilder.java:27]: If not, the relevant calls in PollingMasterInquireClientTest will throw the following exception.
`java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Boolean`

File: core/common/src/main/java/alluxio/master/PollingMasterInquireClient.java (hunk 2 above)
Comment (elega) [reply core/common/src/main/java/alluxio/master/PollingMasterInquireClient.java:159]: ah okay sounds good

**Comment by maobaolong** [comment]
@qian0817 We can use 

```java
          Object value = key.parseValue("xxxx");
          Configuration.set(key, value, Source.RUNTIME);
```

File: core/common/src/test/java/alluxio/master/PollingMasterInquireClientTest.java [hunk 4]
```diff
@@ -46,11 +54,41 @@ public void pollRejectingDoesntHang() throws Exception {
...
     try {
+      s1.start();
+      s2.start();
+      List<InetSocketAddress> addrs =
+          Arrays.asList(InetSocketAddress.createUnresolved("127.0.0.1", port1),
+              InetSocketAddress.createUnresolved("127.0.0.1", port2));
+      PollingMasterInquireClient client = new PollingMasterInquireClient(addrs,
+          () -> new CountingRetry(0),
+          new ConfigurationBuilder()
+              .setProperty(PropertyKey.USER_MASTER_POLLING_CONCURRENT, true)
```
Comment (maobaolong) [inline core/common/src/test/java/alluxio/master/PollingMasterInquireClientTest.java:83]: Object value = key.parseValue("xxxx");
          Configuration.set(key, value, Source.RUNTIME);

Comment (elega) [reply core/common/src/test/java/alluxio/master/PollingMasterInquireClientTest.java:83]: Configuration.set(PROPERTY_KEY, false/true);

**Review by elega — APPROVED** [review]

**Review by elega — APPROVED** [review]

**Comment by elega** [comment]
alluxio-bot, merge this please
